File: app/observability/tracing.py
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

DEFAULT_OTLP_ENDPOINT = "http://localhost:4318/v1/traces"


def configure_tracing(service_name: str, otlp_endpoint: str = DEFAULT_OTLP_ENDPOINT) -> None:
    """Configures the global TracerProvider to export spans over OTLP.

    Call once, from a composition root only (app.main, app.evaluate) —
    never from a module that unit tests import, since tests never call
    this and rely on OpenTelemetry's default no-op tracer instead.
    """
    provider = TracerProvider(resource=Resource.create({"service.name": service_name}))
    provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=otlp_endpoint)))
    trace.set_tracer_provider(provider)

File: services/ingestion/internal/observability/tracing.go
// Package observability configures OpenTelemetry tracing for the
// ingestion CLI. Call Setup once from main, never from a package a unit
// test imports — those never call it and rely on the OpenTelemetry SDK's
// default no-op tracer instead.
package observability

import (
	"context"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

const DefaultOTLPEndpoint = "localhost:4318"

// Setup configures the global TracerProvider to export spans over OTLP.
// The returned shutdown function flushes and closes the exporter; call
// it (deferred) before the process exits.
func Setup(ctx context.Context, serviceName, otlpEndpoint string) (func(context.Context) error, error) {
	exporter, err := otlptracehttp.New(
		ctx,
		otlptracehttp.WithEndpoint(otlpEndpoint),
		otlptracehttp.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}

	res, err := resource.Merge(
		resource.Default(),
		resource.NewSchemaless(semconv.ServiceName(serviceName)),
	)
	if err != nil {
		return nil, err
	}

	provider := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(provider)

	return provider.Shutdown, nil
}
